types/index.d.ts

```diff
@@ -1490,23 +1562,23 @@ export class Provider extends events.EventEmitter {
   prependOnceListener(event: 'server_error', listener: (ctx: KoaContextWithOIDC, err: Error) => void): this;
   // tslint:enable:unified-signatures
 
-  readonly Client: typeof Client;
-  readonly AccessToken: typeof AccessToken;
-  readonly InitialAccessToken: typeof InitialAccessToken;
-  readonly RefreshToken: typeof RefreshToken;
-  readonly AuthorizationCode: typeof AuthorizationCode;
-  readonly RegistrationAccessToken: typeof RegistrationAccessToken;
-  readonly PushedAuthorizationRequest: typeof PushedAuthorizationRequest;
-  readonly ClientCredentials: typeof ClientCredentials;
-  readonly DeviceCode: typeof DeviceCode;
-  readonly BaseToken: typeof BaseToken;
+  readonly Client: ClientConstructor;
```

Comment:
Correct me if i'm wrong but doesn't this change mean that all of [this](https://github.com/panva/node-oidc-provider/pull/838/files?diff=split&w=1#diff-093ad82a25aee498b11febf1cdcb6546e4d223ffcb49ed69cc275ac27ce0ccceR642-R652) and other potentially useful prototype overloadable methods won't be overloadable in TS when accessed through e.g. `provider.Client.prototype.redirectUriAllowed`?

dtto for essentially all the models.

What everyone can do now to get the types, altho not exported, if they need them is, e.g.

```
type Client = InstanceType<Provider['Client']>;
```

Reply:
I am not sure I fully understand. But if you talk about the possibility of doing the inheritance like following or override method, it's always possible:

<img width="660" alt="Capture d’écran 2020-12-14 à 20 57 14" src="https://user-images.githubusercontent.com/1763311/102128995-f98bec00-3e4e-11eb-847c-cfc65567c22e.png">

Unfortunately, I tried to use your example, but it doesn't work:

```typescript
type Client = InstanceType<Provider['Client']>;
``` 
Typescript considers the instanceType as any. In the capture, we can see that the fields aren't available:

<img width="661" alt="Capture d’écran 2020-12-14 à 20 41 00" src="https://user-images.githubusercontent.com/1763311/102127389-bb8dc880-3e4c-11eb-8a7a-3b483787f457.png">

Reply:
I don't care about extending classes, only important thing to me is overloading the prototype methods i consider part of the semver contract. With `provider.Client` being typed as ClientConstructor that only has `find` as a static method all the actual Client instance methods are not typed now.

See the before/after comparison of master/your PR

<img width="808" alt="Screenshot 2020-12-14 at 21 06 05" src="https://user-images.githubusercontent.com/241506/102129871-5dfb7b00-3e50-11eb-9927-7134726d033d.png">
<img width="516" alt="Screenshot 2020-12-14 at 21 06 21" src="https://user-images.githubusercontent.com/241506/102129875-5fc53e80-3e50-11eb-9f3a-5381c906989e.png">

Reply:
And the InstanceType method works just fine.

<img width="415" alt="Screenshot 2020-12-14 at 21 10 25" src="https://user-images.githubusercontent.com/241506/102130217-de21e080-3e50-11eb-87b7-c832d3b204a3.png">

proof: without casting the result as Client.

<img width="823" alt="Screenshot 2020-12-14 at 21 10 35" src="https://user-images.githubusercontent.com/241506/102130248-e7ab4880-3e50-11eb-82f1-63f95137c13b.png">

Reply:
Ok my bad, I realize the problem is on my side. My apologize for the wasted time :) and thanks for Tips ;)
I close the PR now and thanks again.

Reply:
You're welcome and thank you! 

Please consider [supporting](https://github.com/panva/node-oidc-provider?sponsor=1) the library if it provides value to you or your company and this was of help to you. Supporting the library means, amongst other things, that such support will be available in the future.

Reply:
Yes sure, I’m the author of Ts.ED framework and I understand the importance of the support ;). When the new module will be ready, I’ll publish a page to present the oidc-provider and the integration with the framework! I hope the module will help our respective communities ;)

See you